File: leetcode_cpp/convert_sorted_array_to_binary_search_tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:

    TreeNode *sortedArrayToBST_helper(vector<int> &num, int start, int end) {
        if (start > end) return NULL; // important
        int n = (start + end) / 2;
        TreeNode *node = new TreeNode(num[n]);
        node -> left = sortedArrayToBST_helper(num, start, n - 1);
        node -> right = sortedArrayToBST_helper(num, n + 1, end);
        return node;
    }

    TreeNode *sortedArrayToBST(vector<int> &num) {
        return sortedArrayToBST_helper(num, 0, num.size() - 1);
    }
};

File: leetcode_python/convert_sorted_array_to_binary_search_tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param num, a list of integers
    # @return a tree node
    def sortedArrayToBST(self, num):
        if len(num) == 0:
            return None

        n = len(num) / 2
        node = TreeNode(num[n])
        node.left = self.sortedArrayToBST(num[:n])
        node.right = self.sortedArrayToBST(num[n + 1:])
        return node
        
